### ISSUE TITLE
Inspect / Debug when running in sandbox

### ISSUE THREAD

**Issue by darbio** [architect/architect#640]
How do you connect an inspector (e.g. VS code or Chrome dev tools) to functions running in sandbox? I can't find anything in the docs and the normal method (connect to the process) doesn't seem to be hitting breakpoints. Cheers

**Comment by ryanblock** [comment]
Will leave open as help wanted for a bit!

**Comment by ryanblock** [comment]
I think this issue has been open long enough – I remain open to ideas on how to accomplish this, though!

**Comment by darbio** [comment]
@ryanblock one way to do it is already in the thread.

Would have loved to help but I didn’t take enough time working out how to get the sandbox source working locally and my requests for assistance were unanswered. Unfortunately we switched back to our old framework to achieve this. 

This is how the serverless framework does it:

export SLS_DEBUG=* && node --debug ./node_modules/.bin/serverless offline -s dev

I understand that the way arc works is different, however that’s how I would expect it to work (I.e. the “normal” node way).

Child processes that are started by sandbox “just” need to use the child process debug mode I linked to earlier.

**Comment by filmaj** [comment]
I walked through a manual way of attaching a debugger to processes sandbox spawns here in discord: https://discord.com/channels/880272256100601927/884130225280139394/1015995573465260156

It's possible, but requires a few lines of code of change, and careful control of the execution context to avoid the problems Ryan brought up - which, depending on the app, may be difficult.

**Comment by hovissimo** [comment]
I've spent about an hour looking up how to get a debugger attached and then found this thread.  I understand now why `--inspect` isn't already available, but I'll add my use case and note my desire for something like this.

I'm playing around with enhance.dev (which uses architect) to learn more about functional web applications.  I have a very strong desire to attach the debugger so that I can "look around" in the execution context, and evaluate expressions quickly to try and better understand this paradigm that's very new to me.

I'll include a more concrete example. Suppose I have the following function and the first console.log prints, but the second one doesn't.
```javascript
export async function post(req) {                        
  console.log('POST /api/shindigs/$id', {req})
  //validate.create(req) comes from the framework
  let { problems, shindig } = await validate.create(req); 

  console.log({problems, shindig});
  if (problems) {                                        
    // ...                                                
  }                                                      

  // ...
}
```

No exception was thrown, but I definitely got a result I didn't expect.  I want to be able to break after the first console log and step through execution to understand what happened and where I misunderstood the framework.  This can be VERY challenging with logger debugging, but is a great use-case for the debugger protocol.

Re-iterating my point: The debugger is a priceless tool for people who are exploring and don't know where to start when debugging their code, and I suggest Architect should consider something like `--inspect-brk` even if it won't be broadly useful.

**Comment by darbio** [comment]
> Re-iterating my point: The debugger is a priceless tool for people who are exploring and don't know where to start when debugging their code, and I suggest Architect should consider something like `--inspect-brk` even if it won't be broadly useful.

This. 

We would have loved to have used arc, but the lack of support for a critical development tool (a debugger), coupled with the resistance by the author to entertain the addition of this feature, pushed us away and back to serverless.

The suggestion to debug using `console.log` was a step backwards that we just weren't willing to take.

**Comment by ryanblock** [comment]
@darbio re-reading this thread I’m bummed (and surprised) your takeaway was resistance to this feature. Possible since I’m not a frequent user of the debugger I may not fully understand its behavior, but I’m not at all resistant to the addition of this feature.

On that note, both the Architect and Enhance communities have made it clear over the past couple weeks that this feature should be a priority. So it feels appropriate to reopen this issue and get back to work on it!

On my side, as soon as I get the opportunity I’m going to dig back into 
some research and testing, but if anyone would like to spend a little while putting together a summary for someone like me who doesn’t use the buggers very often to address the concern with processors piling up, that would be great!

**Comment by ryanblock** [comment]
Doing some reading this evening, I revisited @jessehattabaugh's clever and elegant idea to add [`debugger` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/debugger) to code to enter Sandbox into debugger / inspection mode.

I think this great in theory, but so far as I can tell, Node only uses `debugger` statements for its own built-in CLI debugging utility, and doesn't pass the statement along as breakpoints to the [V8 inspector integration](https://nodejs.org/api/debugger.html#v8-inspector-integration-for-nodejs) when invoked with `--inspect[-brk]`. Has anyone gotten `debugger` statements to work with V8 debugger connections in Node?

**Comment by ryanblock** [comment]
A quick update, I think this will be feasible to do in a way that doesn't suck, but it is absolutely not trivial, and definitely not as simple as finding a free port and adding the `--inspect-brk` flag to the Lambda execution. Shipping this will be a heavy lift and require some serious beta testing, so I hope y'all will bear with me and pitch in on playing around with some Sandbox builds!

Anyway, here's a local prototype I have of VS Code's debugger attached to a Lambda handler in Sandbox:
<img width="950" alt="image" src="https://user-images.githubusercontent.com/200964/189510108-7946c49b-2667-49e2-be86-80b9a3e43615.png">

If you look closely, you'll see a new global function I'm calling calling: `arcDebug()`. I'm wondering what y'all would think about this method as the affordance of choice for kicking off debugging in your handler?

**Comment by darbio** [comment]
It seems a bit smelly when compared to how this works in other frameworks, where you 'just' connect the debugger to the process and your breakpoints will be hit. 

Why won't adding the `--inspect` flag to the spawned process work?

**Comment by ryanblock** [comment]
> 'just' connect the debugger to the process and your breakpoints will be hit

Which process? I feel like we never really got on the same page about this before. Sandbox's job is to closely emulate key AWS services used in Architect (e.g. API Gateway, Lambda, SNS, etc., to name a few). Just as in Lambda, whenever you invoke the handler, you spin up a new process – it executes, returns (or fails), and exits.

Now assume we've got an actual production Architect app. Some of them have hundreds of Lambdas. begin.com has I think 80-90 unique Lambdas per environment, so we can use that as an example. Let's say I want to debug `POST /foo`, which I need to get to by visiting the home page, logging in, loading a few CSS/JS files from a static asset handler endpoint, then loading whatever view contains the form to `POST` to `/whatever`.

The `--inspect` flag won't work, as you've [correctly established](https://github.com/architect/architect/issues/640#issuecomment-544238023) – we're emulating Lambda; since it's not a long-lived process, we have to break before user code starts, and thus must use `--inspect-brk` or something like it. If I "just" turn on debugging add the `--inspect-brk` flag, assuming I actually know the order of debug instances to attach to and execute, I would have to attach to dozens before I can finally get to the one handler I'm actually trying to debug. No one would do that, it would be ridiculous.

tldr debugging _must_ break before user code starts (or the Lambda will simply run to completion and exit before the developer has a chance to attach to it), and the developer _must_ identify which handler(s) they'd like to debug (otherwise apps with multiple functions will be impossible to debug).

The approach I've outlined above is vaguely similar to employing the `debugger` keyword (and actually supports `debugger` statements, which by the way I believe `--inspect[-brk]` does not).

That said, I'm open to other ideas – and PRs – for solving this so long as they account for the above considerations.

> Could you bake in the call to arcDebug to the framework, perhaps only calling when NODE_ENV is not production?

Not really – Architect, as a framework, is generally lower-level than this, and we don't ever mutate userland code; putting something like an `arcDebug` method into global scope is totally a special case to enable individual handlers to opt into Sandbox debugging.

So if one were to let an `arcDebug()` statement into live Lambda code, it will blow up and fail, because it simply won't be in global. You could in theory mitigate that by doing something like `if (arcDebug) arcDebug()`, but that would be up to the developer.

**Comment by darbio** [comment]
I don't think anyone would be attaching a debugger to a production app (very bad idea). The debugging feature would (should?) be only on your local dev box. 

Given this assumption, sandbox could allow itself to be started with a `--debug` (or `--singleprocess`) flag and spin up each handler in the same process, instead of an individual process.

Another approach is for sandbox to allow all handlers to run on one process. 

That way you just attach the debugger to the sandbox process and all debugger breakpoints will be hit. 

Yes, this differs from the way that production code is executed, but in the sandbox does that really have much of an impact? It's not like sandbox is used as a production web server anywhere.

**Comment by ryanblock** [comment]
Since I don’t seem to be hitting the mark in my explanations of how Lambda and Sandbox both work, please take a few to check out the Sandbox codebase and feel free to send a PR if you feel like the approach you’re suggesting is feasible.

**Comment by darbio** [comment]
I'm going to bow out here. I don't use your framework and I'm trying to be constructive in my suggestions, but it's obvious that they are not welcome. 

I did look into the way that sandbox worked back when I asked this question, and I am well aware of how lambda works having multiple enterprise applications in production, running on AWS lambda for nearly a decade. 

I asked a simple question as to how to attach a debugger to code being run in sandbox, something that isn't eccentric to expect to be able to do. My request was answered with a trite response of "use console.log", which was quite frankly unhelpful.

Good luck with implementing this feature as I believe it will definitely be a killer feature that will help in obtaining and keeping developers.

**Comment by filmaj** [comment]
@darbio did you see my [earlier comment](https://github.com/architect/architect/issues/640#issuecomment-1241451284) on how to get it working? I walked through how to set it up and what 2 lines of code need changing.

It definitely does not address the valid concerns that Ryan brings up, and as a past contributor to this project, I as well wouldn't feel comfortable steaming ahead and adding first-class support for attaching a debugger to sandbox-spawned processes without thinking through how this could cause a cascade of failures and a poor experience in an app that is slightly more complex than a hello world sample.

> I'm going to bow out here. I don't use your framework and I'm trying to be constructive in my suggestions, but it's obvious that they are not welcome.

It's a great suggestion, and worth investigating / hacking / pushing the boundaries to see what is possible. However, and with all due respect, talk is cheap, while work and PRs and writing tests and getting the feature to pass in the plethora of different use cases that architect supports is much harder. From my perspective, a few people in this thread have offered different approaches (Ryan even spent a non-trivial amount of his time researching one way of doing it, and asking for feedback on that) which it didn't look like to me there was any recognition of or constructive conversation around.

**Comment by ryanblock** [comment]
> My request was answered with a trite response of "use console.log", which was quite frankly unhelpful.

I never said those words, or anything like them. You asked how I debug, [I replied](https://github.com/architect/architect/issues/640#issuecomment-544264843) “console.log works pretty well!” How is that unhelpful? That is the true answer to that question (and the answer for many JS devs). It’s not flip, it’s simply how I debug code.

@darbio your end of this conversation feels a bit bad faith. I wish you the best of luck, and do appreciate your calling attention to this functionality gap. One way or another, it will soon be addressed!

**Comment by darbio** [comment]
> your end of this conversation feels a bit bad faith

No bad faith here, just frustration. 

> @darbio did you see my earlier comment on how to get it working? I walked through how to set it up and what 2 lines of code need changing.

@filmaj yes, I saw your comment but didn't appear to have access to the discord channel. Can you post here for posterity?

The use case I wanted was:

As a developer I want to be able to attach to the framework so that I can inspect the execution of the code. 

The way that I do this in other frameworks is to attach to the process, or to start the process from an IDE such as VS Code. 

This framework shouldn't require anything special **from an end user point of view** to make it work - attach to the process and your breakpoints should be hit. 

The proposed solution using `arcDebug` is novel (special) from an end user point of view and not how an end-user would expect inspection to be implemented.

**Comment by ryanblock** [comment]
I definitely sense your frustration – it seems that underlying your feedback is an apparent desire for Architect to adopt a model of utilizing Lambda that it eschews. That model is sometimes called the ["fat Lambda"](https://hackernoon.com/why-you-should-stop-using-fat-functions-in-your-lambda-code) – Architect takes an alternative approach, driving folks towards adherence to the [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle), and to conform to the [Functional Web App model](https://fwa.dev), and that approach is utterly fundamental.

Other frameworks are ok with (and even encourage) fat Lambdas at the cost of performance, debuggability, and long-term maintainability. Architect is designed from the ground up, in every possible way, to utilize many small, single responsibility Lambdas, each running in isolation.

To that end, while it may theoretically be technically possible for Sandbox to load all various your Lambdas' handlers into a single running process, such an approach would be a total rewrite, and would be fundamental break with how Lambda works (e.g. handlers are isolated; no long-lived processes).

That doesn't mean we can't do V8 inspection, but it just won't look _exactly_ as you're used to seeing it.

> The way that I do this in other frameworks is to attach to the process, or to start the process from an IDE such as VS Code.

I think we can get relatively close to that. What I proposed above would look like this: start Sandbox > fire your handler (e.g. load `get /` where `src/http/get-index/index.js` contains a debug statement of some kind), attach your inspector. In VS Code, attaching would mean having a 5 line `launch.json` file, and selecting the handler you want to attach your debugger to once it's been fired.

> This framework shouldn't require anything special from an end user point of view to make it work - attach to the process and your breakpoints should be hit.

I generally agree; see above, I think we can get pretty close.

> The proposed solution using arcDebug is novel (special, not standard) from an end user point of view and not how an end-user would expect inspection to be implemented.

That's totally fair, and why I asked for feedback. The same can also be said of debugging Lambda in general. There's no staging server with a debug port running for you to connect to, after all. So you can also choose to look at this a different way: if you want to debug your code in Lambda, you're likely using Xray or CloudWatch logs. Those could be said to be "special" and "not standard" and "not how an end-user would expect inspection to be implemented", and yet that's how the majority of folks debug Lambda (which is now the largest compute primitive in the largest cloud provider). And as it happens, that debugging model transfers nicely over to Architect: with a V8 inspector and `console.log`, one Lambda connection at a time.

All that said, to echo Fil's point, code speaks louder than words. I'm out here hacking on a PR – if you (or anyone) have alternate ideas for more elegant approaches I'd be so stoked to see them!

**Comment by darbio** [comment]
I _think_ were arguing for a similar thing, slight nuances, but neither of us are _getting_ the other. 

I wish we could have this discussion in person. So much of the art of communication is missed when it's just text. 

>  it seems that underlying your feedback is an apparent desire for Architect to adopt a model of utilizing Lambda that it eschews. That model is sometimes called the "fat Lambda"

Absolutely not - fat lambda is an anti pattern. I'm not advocating this at all. 

And I agree that each handler should adhere to SRP.

I'm not trying to eschew sound engineering practices. At all. 

My argument is as follows (and it's not meant to rubbish the way sandbox currently works):

Sandbox is not meant to be run as a production server, I am working on the assumption that it is supposed to be for dev box only work (and some unit/integration testing). 

There are other ways to emulate the way that lambda works (i.e. handlers are isolated, no long-lived processes etc.), and I feel that on a dev box, one of the major requirements is for a great developer experience, including being able to inspect the running handler.

Emulating the actual way that something is implemented is obviously important, but doesn't have to be done to the letter. 

From my initial research 3 years ago I remember that sandbox spawns a new process to run the handler, which is what causes the issue of not being able to attach a debugger.

So in order to attach the debugger, either the lambda needs to be run in the same process, or there needs to be a way to determine which process to connect to without having to use `debugger` or `--inspect-brk `.

Running in the same process allows the whole thing to "just" work, as in right now. Obviously the process level isolation is lost, but how essential is that on a dev box? This is a change to the way that sandbox currently works. 

If you want to maintain the current process level isolation, then I don't see a simple solution other than spinning up new child processes with the `--inspect` flag (actually, `--inspect-brk` as the process is killed after execution so there isn't enough time to connect). This is the same as `arcDebug` and it's not the _standard_ way of debugging code which makes me think of it as a code smell. 

> if you want to debug your code in Lambda, you're likely using Xray or CloudWatch logs. Those could be said to be "special" and "not standard" and "not how an end-user would expect inspection to be implemented", and yet that's how the majority of folks debug Lambda

I don't expect the framework to be able to even connect to AWS Lambda as that is not a sandbox concern at all. 

> code speaks louder than words

I know... and I wish I had more time to look into this with you.

**Comment by filmaj** [comment]
> but didn't appear to have access to the discord channel. 

Ah, odd, here it is slightly edited for GitHub:

> for the purposes of this exercise i just edited sandbox code direclty under `node_modules/@architect/sandbox` of my app, just to quickly hack around and see if any of this is at all possible; good luck!

1. i added a `--inspect-brk` flag before the `-e` [here](https://github.com/architect/sandbox/blob/main/src/lib/runtime-eval.js#L15). for some reason just the `--inspect` flag + adding a `debugger` statement in my code didn't work for me - the chrome debugger wouldn't stop at the debugger statement 🤷‍♂️ (sounds like Ryan has looked into this somewhat)
2. NB: attaching a debugger to any of your lambda routes will probably make them timeout. Arc has default timeouts for many lambdas, and this can further be customized in each function's folder via the[ function `timeout` config](https://arc.codes/docs/en/reference/configuration/function-config#timeout). i overwrote the timeout parameter in [this line](https://github.com/architect/sandbox/blob/main/src/invoke-lambda/exec/spawn.js#L13) to set it to some ridiculously high number. possibly a different approach here would be set the `timeout` function config to the max value and hope that works in a debugger context (900 seems to be the max)
3. started `arc sandbox`. once i hit an http route (i.e. http://localhost:3333/) , i could see in the output of my sandbox process something like: 

    Debugger listening on ws://127.0.0.1:9229/b9df8328-9570-4542-809c-bfc68e3ce16b
    For help, see: https://nodejs.org/en/docs/inspector

4. load up the chrome inspector: chrome://inspect
5. under 'Remote Target' i could see a node process was available. click that, it opens a new chrome inspector window with the arc sandbox code
<img width="419" alt="Screen_Shot_2022-09-04_at_12 24 26" src="https://user-images.githubusercontent.com/52645/189569146-5d4abf71-9514-45ce-b33f-aa156b744b5b.png">
6. sandbox's node runtime 'bootstrap' script - the one that orchestrates invoking your userland function , which it has one for each runtime arc supports - is minified, i clicked the 'pretty-print' curly braces button in the bottom left corner of the window to un-minify and make it easier to step through
<img width="1036" alt="Screen_Shot_2022-09-04_at_12 25 14" src="https://user-images.githubusercontent.com/52645/189569242-b07609bd-644d-4295-9090-ba69827fed93.png">
7. scroll all the way down to where your function gets invoked (line 114 in my case) and pop a breakpoint there. hit the play button to get to that line. click the Step Into button
<img width="895" alt="Screen_Shot_2022-09-04_at_12 27 33" src="https://user-images.githubusercontent.com/52645/189569405-929972f2-a638-4ac5-b036-a576b3a20e56.png">
  - for some reason theres this weird promise related code that comes up when i step into. i Stepped Over this section three times to get through it
<img width="642" alt="Screen_Shot_2022-09-04_at_12 28 17" src="https://user-images.githubusercontent.com/52645/189569437-64689bee-4026-464a-9095-0f2002698637.png">
  - since my userland code uses the optional `@architect/functions` http async helper, I also had to step through that

as was discussed here before, ideally if it could be figured out how to properly use node's `--inspect` flag with `debugger`, that'd be the best experience. but i didnt have time to explore why that wasn't working for me. that would save many of the above steps of stepping through framework or native code and let you fast-forward straight to your own app's code.

**Comment by ryanblock** [comment]
The good news is: fast forwarding to the userland handler code is what I’ve got solved, and I may be able to do it without that global method (stand by…). Other problems still remain to be solved, though, like:
- What to do when a handler being inspected is called again before the inspector is detached? (I am thinking all additional requests to a handler being inspected would fail until the handler completes invocation and inspection.)
- How to identify in the console and VS Code which process / invocation equates to which handler when there are multiple handlers available to attach to
- How to detect attached inspectors and terminate them if the user quits Sandbox. (This one may wind up being the trickiest to solve, tbqh.)

**Comment by darbio** [comment]
> What to do when a handler being inspected is called again before the inspector is detached? 

Normal behaviour would be to block execution until the handler has exited. Once it has executed, other code can be hit (handlers called etc.). 

If you allow more calls, you run the risk of stepping through two processes at the same time which is confusing. 

> How to identify in the console and VS Code which process / invocation equates to which handler when there are multiple handlers available to attach to

You could do this through having the handler run on the same port (e.g. re-use the port). 

> How to detect attached inspectors and terminate them if the user quits Sandbox

I think this is a non issue. The debugger will disconnect when the process is killed. Are child processes killed when the sandbox process is killed?

I still think that this flow is not the best developer experience as you need explicit knowledge of how to debug, rather than just attaching a debugger to the sandbox process - but I am keen to see how your implementation works.

**Comment by darbio** [comment]
An example of running the lambda offline in the same process can be found in [serverless offline](https://github.com/dherault/serverless-offline/blob/master/src/lambda/handler-runner/in-process-runner/InProcessRunner.js).

This method would break your model of having a process per lambda, but would allow you to attach the debugger to the sandbox process rather than the lambda process.

It's still isolated in the sense that sandbox calls the handler with the input and can't inject anything else. 

Fundamental change but I don't think it breaks anything in the rest of your code base?

Maybe this in process runner could be added as an option so you maintain the separation of processes for unit / integration testing, but can choose to run in process if you require the debugger?

**Comment by jessehattabaugh** [comment]
@ryanblock 

>  I’m bummed (and surprised) your takeaway was resistance to this feature. Possible since I’m not a frequent user of the debugger I may not fully understand its behavior, but I’m not at all resistant to the addition of this feature.

Resistance to complexity is a valuable instinct in software. I wouldn't fault you for it even if it were true! 😄

**Comment by fuzzybaird** [comment]
Hey, I just wanted to add to this. I appreciate everyone's time looking into this and gauging the scope of implementation and the debates around SRP. I tend to agree with @darbio, that when it comes to local environments, devs are mostly looking to understand their code, and its lifecycles. Flagging the debugger on and off when releasing to production seems like a compromise that meets everyone's needs.

The scope of work concern is valid, but it is a pretty used feature at most dev shops I have worked for, especially with node backgrounds. That could weigh into the value vs. effort decision. I also might be willing to chip in and try and make a PR if there is an appetite, I am just not the best node guy, and would need a little guidance.

**Comment by ryanblock** [comment]
Last year I put a number of days into trying to figure out how to make this work without fundamentally altering the compute model of Sandbox. I got pretty far (as in: I got the debugger attaching to individual processes while executing with a bit of finagling), but ultimately a number of blockers remained and I ran out of time. Defeat!

I stashed my changes last September with the intention to get back to them. Well, that was ~123 commits back, by my estimation.

In the interest of science, I've resurrected that stash as a new PR, with the following caveats:
- When I last left off, I was able to attach a debugger in VS Code, but it was wonky as hell, especially with multiple Lambdas executing at the same time (which is often the case)
- Since I authored this stash, Sandbox has seen some pretty radical changes under the hood, especially with the implementation of the Lambda Runtime API; please refer to [everything above this](https://github.com/architect/sandbox/blob/main/changelog.md#534-2022-09-07)
- Given drift over the last 123 commits, this code may no longer be working. In fact, if you're going to check it out, it's probably best to assume it doesn't work anymore
- The may not even be the best or most correct approach – in other words, the PR may be a complete blind alley
- I barely remember how this works, and I cannot guarantee I even merged it correctly, so support will not be offered 🫠

Here you go! https://github.com/architect/sandbox/pull/719

**Comment by quantuminformation** [comment]
thanks again, this is pretty much essential for our workflow
